src/storage/rect.rs

```diff
@@ -0,0 +1,632 @@
+    ///     .with_half_size(IVec2 { x: 8, y: 4 })
+    ///     .with_wrap_strategies([WrapStrategy::Cycle, WrapStrategy::Clamp])
+    ///     .build_default::<i32>();
+    ///
+    /// assert_eq!(rect_map.get(Hex::new(0, 0)), Some(&0));
+    /// ```
+    #[must_use]
+    pub fn build_default<T: Default>(self) -> RectMap<T> {
+        RectMap::default_values(self)
+    }
+
+    // ================================
+    // Access
+    // ================================
+    /// get the offset mode of the map
+    #[must_use]
+    pub const fn offset_mode(&self) -> OffsetHexMode {
+        self.offset_mode
+    }
+    /// get the half size of the map
+    #[must_use]
+    pub const fn half_size(&self) -> IVec2 {
+        self.half_size
+    }
+    /// get the wrap strategies of the map
+    #[must_use]
+    pub const fn wrap_strategies(&self) -> [WrapStrategy; 2] {
+        self.wrap_strategies
+    }
+
+    // ================================
+    // Forward Coordinate Conversion
+    // ================================
+    /// calculate hex coordinate from index
+    ///
+    /// infallible
+    ///
+    /// - `idx` `Vec` index
+    /// - => `rc` 2D view of `Vec`
+    /// - => `ij` offset coordinate
+    /// - => `hex`
+    #[allow(clippy::cast_possible_truncation, clippy::cast_possible_wrap)]
+    fn idx_to_hex(&self, idx: usize) -> Hex {
+        let idx = idx as i32;
+        let rc = IVec2::new(idx % (2 * self.half_size[0]), idx / (2 * self.half_size[0]));
+        let ij = rc - self.half_size;
+        self.ij_to_hex(ij)
+    }
+
+    fn ij_to_hex(&self, ij: IVec2) -> Hex {
+        Hex::from_offset_coordinates(
+            ij.xy().to_array(),
+            self.offset_mode,
+            self.hex_layout.orientation,
+        )
+    }
+
+    // ================================
+    // Backward Coordinate Conversion
+    // ================================
+    /// `None` if outside map
+    fn hex_to_idx(&self, hex: Hex) -> Option<usize> {
+        let ij = self.hex_to_ij(hex);
+        if self.contains_ij(ij) {
+            Some(self.ij_to_idx(ij))
+        } else {
+            None
+        }
+    }
+
+    /// Wrap if `hex` lie outside of layout
+    fn wrapped_hex_to_idx(&self, hex: Hex) -> usize {
+        let ij = self.hex_to_ij(hex);
+        let ij = self.wrap_ij(ij);
+        self.ij_to_idx(ij)
+    }
+
+    /// infallable
+    fn hex_to_ij(&self, hex: Hex) -> IVec2 {
+        let v: IVec2 = hex
+            .to_offset_coordinates(self.offset_mode, self.hex_layout.orientation)
+            .into();
+        v.xy()
+    }
+
+    /// fallable input, internal
+    ///
+    /// - `ij` offset coordinate
+    /// - => `rc` 2D view of `Vec`
+    /// - => `idx` `Vec` index
+    fn ij_to_idx(&self, ij: IVec2) -> usize {
+        let rc = (ij + self.half_size).as_uvec2();
+        (rc[0] + rc[1] * (2 * self.half_size.as_uvec2()[0])) as usize
+    }
+
+    /// infallable
+    fn wrap_ij(&self, mut ij: IVec2) -> IVec2 {
```

Comment:
Please use `.x` and `.y` fields instead of indexing on the `IVec2`, those would add clarity and reduce the amount of faillible operations